Control/Monad/Logic/Class.hs

```diff
@@ -45,19 +45,56 @@ class (MonadPlus m) => MonadLogic m where
     --   > do x <- odds `mplus` return 2
     --   >    if even x then return x else mzero
     --
-    --   Such a computation may never consider the 'return 2', and will
-    --   therefore never terminate. By contrast, interleave ensures fair
-    --   consideration of both branches of a disjunction
+    --   Such a computation may never consider the @return 2@, and
+    --   will therefore never return any results. By contrast, using
+    --   'interleave' in place of 'mplus' ensures fair consideration
+    --   of both branches of a disjunction.
+    --
+    --   Note that this sample computation will never terminate: only
+    --   the first value can be safely observed, after which each odd
+    --   value becomes 'mzero' (equivalent to a Prologue @fail@) which
+    --   does not stop the evaluation but indicates there is no value
+    --   to return yet.
+    --
     interleave :: m a -> m a -> m a
 
     -- | Fair conjunction. Similarly to the previous function, consider
-    --   the distributivity law for MonadPlus:
+    --   the distributivity law for 'MonadPlus':
     --
     --   > (mplus a b) >>= k = (a >>= k) `mplus` (b >>= k)
     --
-    --   If 'a >>= k' can backtrack arbitrarily many times, (b >>= k) may never
-    --   be considered. (>>-) takes similar care to consider both branches of
-    --   a disjunctive computation.
+    --   If @(a >>= k)@ can backtrack arbitrarily many times, @(b >>=
+    --   k)@ may never be considered. The '>>-' operator takes similar
+    --   care to consider both branches of a disjunctive computation.
+    --
+    --   Consider the operation:
+    --
+    --   > odds = return 1 `mplus` liftM (2+) odds
+    --   >
+    --   > oddsPlus n = odds >>= \a -> return (a + n)
+    --   >
+    --   > do x <- (return 0 `mplus` return 1) >>= oddsPlus
+    --   >    if even x then return x else mzero
+    --
+    --   This will never produce any values because all values come
+    --   from the @return 1@ driven operation, but the @return 0@ driven
+    --   operation generates an infinite number of potential results.
+    --   Using 'interleave' here instead of 'mplus' does not help due
+    --   to the MonadPlus distributivity law.
+    --
+    --   Also note that the @do@ notation desugars to '>>=' bind
+    --   operations, so the following would also fail:
+    --
+    --   > do a <- return 0 `mplus` return 1
+    --   >    x <- oddsPlus a
+    --   >    if even x then return x else mzero
```

Comment:
Interestingly, I played a bit with the `RebindableSyntax` pragma to see if >>- could be automated in `do` notation, but while >>- is declared as the implementation of >>=, the default instance of >>- doesn't seem to be applied.  I suspect this might need a `RULES` specification, but that would necessitate the removal of `Safe` as well as the use of optimization to be *correct*, and that didn't seem desireable (even if you had a situation where `RebindableSyntax` could be considered desireable).

Reply:
> the default instance of >>- doesn't seem to be applied. 

Could you please elaborate? I thought it should work fine, when writing https://github.com/Bodigrim/logict/issues/8#issuecomment-740146446, but I might have been tricked by GHC (because all lists in that example are finite, one could not tell `>>-` from `>>=`).

Another option for automated `do` notation is [`QualifiedDo`](https://downloads.haskell.org/ghc/9.0.1-rc1/docs/html/users_guide/exts/qualified_do.html). Could you possibly check if it works?

Reply:
I'll create another PR with some of this work; perhaps I just did something wrong and you can help fix it.